mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -55,11 +52,207 @@ def _get_optimizer_name(optimizer):
+
+            self.metrics_logger.record_metrics(metrics, pl_module.current_epoch)
+
+    _pl_version = Version(pl.__version__)
+
+    # In pytorch-lightning >= 1.4.0, validation is run inside the training epoch and
+    # `trainer.callback_metrics` contains both training and validation metrics of the
+    # current training epoch when `on_train_epoch_end` is called:
+    # https://github.com/PyTorchLightning/pytorch-lightning/pull/7357
+    if _pl_version >= Version("1.4.0dev"):
+
+        @rank_zero_only
+        def on_train_epoch_end(
+            self, trainer, pl_module, *args
+        ):  # pylint: disable=signature-differs,arguments-differ,unused-argument
+            self._log_metrics(trainer, pl_module)
+
+    # In pytorch-lightning >= 1.2.0, logging metrics in `on_epoch_end` results in duplicate
+    # metrics records because `on_epoch_end` is called after both train and validation
+    # epochs (related PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/5986)
+    # As a workaround, use `on_train_epoch_end` and `on_validation_epoch_end` instead
+    # in pytorch-lightning >= 1.2.0.
+    elif _pl_version >= Version("1.2.0"):
+
+        # NB: Override `on_train_epoch_end` with an additional `*args` parameter for
+        # compatibility with versions of pytorch-lightning <= 1.2.0, which required an
+        # `outputs` argument that was not used and is no longer defined in
+        # pytorch-lightning >= 1.3.0
+
+        @rank_zero_only
+        def on_train_epoch_end(
+            self, trainer, pl_module, *args
+        ):  # pylint: disable=signature-differs,arguments-differ,unused-argument
+            """
+            Log loss and other metrics values after each train epoch
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            # If validation loop is enabled (meaning `validation_step` is overridden),
+            # log metrics in `on_validaion_epoch_end` to avoid logging the same metrics
+            # records twice
+            if trainer.disable_validation:
+                self._log_metrics(trainer, pl_module)
+
+        @rank_zero_only
+        def on_validation_epoch_end(self, trainer, pl_module):
+            """
+            Log loss and other metrics values after each validation epoch
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            self._log_metrics(trainer, pl_module)
+
+    else:
+
+        @rank_zero_only
+        def on_epoch_end(self, trainer, pl_module):
+            """
+            Log loss and other metrics values after each epoch
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            self._log_metrics(trainer, pl_module)
+
+    @rank_zero_only
+    def on_train_start(self, trainer, pl_module):
```

Comment:
I think this is a proper usage for `rank_zero_only`. Can we remove `rank_zero_only` applied on `patched_fit`?

Reply:
Ref: Implementation of `rank_zero_only`:

https://github.com/PyTorchLightning/pytorch-lightning/blob/4928dc55790eec4ee4a558f494a6fc3d49a5e78a/pytorch_lightning/utilities/distributed.py#L44-L65

```python
def rank_zero_only(fn: Callable) -> Callable:
    @wraps(fn)
    def wrapped_fn(*args: Any, **kwargs: Any) -> Optional[Any]:
        if rank_zero_only.rank == 0:
            return fn(*args, **kwargs)
        return None

    return wrapped_fn


# TODO: this should be part of the cluster environment
def _get_rank() -> int:
    rank_keys = ("RANK", "SLURM_PROCID", "LOCAL_RANK")
    for key in rank_keys:
        rank = os.environ.get(key)
        if rank is not None:
            return int(rank)
    return 0


# add the attribute to the function but don't overwrite in case Trainer has already set it
rank_zero_only.rank = getattr(rank_zero_only, "rank", _get_rank())
```